// @ng-matero/extensions/esm2020/core/pipes/pipes.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MtxToObservablePipe } from './to-observable.pipe';
import * as i0 from "@angular/core";
export class MtxPipesModule {
}
/** @nocollapse */ MtxPipesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: MtxPipesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ MtxPipesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.2", ngImport: i0, type: MtxPipesModule, declarations: [MtxToObservablePipe], imports: [CommonModule], exports: [MtxToObservablePipe] });
/** @nocollapse */ MtxPipesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: MtxPipesModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.2", ngImport: i0, type: MtxPipesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [MtxToObservablePipe],
                    declarations: [MtxToObservablePipe],
                }]
        }] });
